Create functional tests #45

Farcasut opened the pull request.

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I added the pom.xmls

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I added the pom.xmls

Farcasut committed:
Added the structure for the config

Farcasut committed:
Implmentation

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
*I added the classes need to interact with the configs from Analytics.Code.COnfiguration and I also added a class that will normalise the jsons from matomo

Farcasut committed:
Merge branch 'main' into livetable

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I added the scripting service and the MatomoAnalyticsManager

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
*I added the livetable and everything seems to work okay for the time being.

Farcasut committed:
[MISC FIX] Indentation

Farcasut committed:
[MISC FIX] Comments

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I modified the implementations of the JsonNormaliser to be a component and I also added the url + the document reference to the returned json.

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I restructured the packages and replaced the URLBuilder with a proper one.

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I added some comments. The final version of the pr.

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I added an if in the case that I switch back to the getPageTitles API

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I modified the dependencies and removed some parameters.

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I modified how I get the normaliser

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I added 2 tests and modified a few typos.

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I added some functions.

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* Renamed the MostViewed hint to MostViewedPages and modified the comments

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* Renamed 2 pages and added a test class for MatomoAnalyticsManager

Farcasut committed:
[MISC FIX]
*Changed the name of a file

Farcasut committed:
Merge branch 'main' into livetable

Farcasut committed:
Removed a file

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I added a test for the MatomoANalyticsManager

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I added new error messages and I updated the comments ona few functions.

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10 * Removed a template

Farcasut committed:
Row Evolution Feature for the Most Viewed Pages #15
* I added the JSON normalizer to transform the Matomo result into a format that can be easily processed in JavaScript for the Chart.js library. I also added two new Velocity macros and the JS code for displaying the evolution.

Farcasut committed:
Row Evolution Feature for the Most Viewed Pages #15
* I added a test for the RowEvolutionJsonNormaliser and added comments to the javascript code.

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* Removed unused variables

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* Removed unused livetable fields

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* Moved the code that checks the date to macro to be able to reuse it in other pages

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* I modified the javascript code and the macros code to be able to reuse the code in other places

Farcasut committed:
Delete RowEvolution.xml

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* Changed the name of a macro

Farcasut committed:
Merge remote-tracking branch 'origin/RowEvolution' into RowEvolution

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* Missing indentation

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* I added a file that I deleted by mistake

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I updated the MatomoANalyticsMangerTest class

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I modified the comments from the JsonNormaliser and added a waring for the developer for when the normaliser hint is not valid

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I added a test for when the Normaliser hint is invalid

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* Update test

Farcasut committed:
Row Evolution feature for the MostViewedPages xwikisas#15
* Updated the javascript code to remove some bugs

Farcasut committed:
Merge branch 'livetable' into RowEvolution

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* Modified the structure of the files

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* Modified the structure of the files

Farcasut committed:
Merge branch 'livetable' into RowEvolution

Farcasut committed:
Row Evolution feature for the MostViewedPages xwikisas#15
* Updated the structure.

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
 * I removed a field form the live table and updated the naming to avoid duplicates

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* I removed a field form the livetable and modified how the title is processed.

Farcasut committed:
Merge branch 'livetable' of https://github.com/Farcasut/application-analytics into livetable

Farcasut committed:
[MISC] Merged livetable branch

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* Modified some comments and fixed the indentation

Farcasut committed:
Merge branch 'site_search_keyword' into demo

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* Modified the classes used for the http request to be compatible with java 8.

Farcasut committed:
Merge branch 'livetable' into RowEvolution

Farcasut committed:
Move Most Viewed Pages Macro to livetable #10
* [MISC] Typo

Farcasut committed:
Row Evolution feature for the MostViewedPages xwikisas#15
* Bug fix

Farcasut committed:
Merge branch 'livetable' into RowEvolution

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* Merged main

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* [MISC] Typo

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* Added translations for all the fields

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* [MISC] Typo

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* Added some translations

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* Modified the tracking code and added new macros to create the modal

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* Renamed the RowEvolutionMostViewedPagesJSON.xml to RowEvolutionJson.xml and solved the request url problem.

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* Removed an ID that was left and solved a bug

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* [MISC] removed an id

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* I added comments and modified the json page to be more general

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* Added a loading animation and solved a bug where the period wouldn't reset.

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* Modified the name of the css classes.

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* [MISC] Updated the js

Farcasut committed:
Moved the MostViwedMacro to livedata from livetable

Farcasut committed:
Migrate MostViewedMacro from livetable to livedata #21
* [MISC] Missing spaces

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
*[Misc]

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* Added the chartjs application as a dependency.

Farcasut committed:
Bug Fixes:
-Fixed LiveData pagination.
-Improved LiveData filtering. Added a method to filter text based on user input. For now, title filtering searches the URL, not the page title.
-Fixed LiveData sorting issues by removing $request.reqNo != '1'.

New Features:
-Changed title property to show both page title and link to the page.

Farcasut committed:
[MISC] Changed the name of a function

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
 * -Added ids to the labels for the rowEvolution modal
 - Added a new parameter for the line color

Farcasut committed:
Merge branch 'main' into migrateToLivedata

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* Merged the MostViewedMacro and removed the search parameter from the livedata because it was redundant.

Farcasut committed:
Migrate MostViewedMacro from livetable to livedata #21
* Added a method to process the nodes and added a new test for the filtering.

Farcasut committed:
Migrate MostViewedMacro from livetable to livedata #21
* Removed an unused variable.

Farcasut committed:
Merge branch 'migrateToLivedata' into RowEvolution

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
* Added a new file to store the jsons for the tests

Farcasut committed:
Migrate MostViewedMacro from livetable to livedata #21
* Added new tests and a new macro to get the subset of a json

Farcasut committed:
Merge branch 'migrateToLivedata' into RowEvolution

Farcasut committed:
Migrate MostViewedMacro from livetable to livedata #21
- Fixed a bug where the subSet macro would return the wrong number of elements

Farcasut committed:
Merge branch 'migrateToLivedata' into RowEvolution

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
- Removed the ability to sort the actions column

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
- Added a filtering system for the Row Evolution Feature

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
- Adding new tests

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
- Removed a variable

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
 - Merged the main branch

Farcasut committed:
Row Evolution feature for the MostViewedPages #15
 - Merged the main branch

Farcasut committed:
Test functiona

Farcasut committed:
Merge branch 'main' into functionalTests

Farcasut committed:
test

Farcasut committed:
page object

Farcasut committed:
test

Farcasut committed:
functional tests

Farcasut committed:
Check the matomo credentials for the api

Farcasut committed:
env for ExtraTools

Farcasut committed:
test
 *
- Added escapetool.xml to the translations

Farcasut committed:
Test container nou

Farcasut committed:
Merge branch 'main' into functionalTests

Farcasut committed:
dashboard dep

Farcasut committed:
Merge branch 'main' into functionalTests

Farcasut committed:
Merge branch 'main' into functionalTests

Farcasut committed:
test

Farcasut committed:
Functional tests

Farcasut committed:
Draft Functional Tests #38
 *
- First functional test(Checks the ability of an admin user to add and remove a macro from the main page)
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

Farcasut committed:
Merge branch 'main' into functionalTests

Farcasut committed:
Admin PageView

Farcasut committed:
*
Added 2 new tests one that checks wrong configs and one checks valid configs.

Farcasut committed:
Matomo page object
* Dosen't work for the time being I can't connect to the matomo application from the testing container.

Farcasut committed:
Merge branch 'main' into functionalTests

Farcasut committed:
Backup

Farcasut committed:
BACKUP

Farcasut committed:
Merge branch 'main' into functionalTests

Farcasut committed:
Backup

Farcasut committed:
Create functional tests #45
* Test for the description button

Farcasut committed:
Merge branch 'main' into functionalTests

Farcasut committed:
Create functional tests #45
* BACKUP

Farcasut committed:
Create functional tests #45
* Added a few functional tests for the basic operations of the extension.

Farcasut committed:
Create functional tests #45
* Removed a file

mflorea commented on application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolutionJson.xml:
```diff
@@ -0,0 +1,54 @@
+  'method' : 'Actions.getPageUrl',
+  'format' : 'json'
+})
+#if ($request.macroName=='MostViewedPages')
```
```suggestion
#if ($request.macroName == 'MostViewedPages')
```

mflorea commented on application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolutionJson.xml:
```diff
@@ -0,0 +1,54 @@
+  $parameters.put('pageUrl', $request.parameter)
+#end
+#set ($analyticsResult = $services.analytics.getMatomoRequestResult($parameters, {}, 'RowEvolution'))
+#jsonResponse($analyticsResult)
```
Would be good to wrap this code with an IF checking for ``get`` action, otherwise it's a pain to interact with this page (edit, view history, etc.).

mflorea commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/MostViewedMacroViewPages.java:
Please format code using XWiki's code style. See https://dev.xwiki.org/xwiki/bin/view/Community/CodeStyle/JavaCodeStyle/#HConfiguringyourIDEtousetheXWikicodestyle .

mflorea commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/MostViewedMacroViewPages.java:
```diff
@@ -0,0 +1,71 @@
+import org.xwiki.test.ui.XWikiWebDriver;
+import org.xwiki.test.ui.po.ViewPage;
+
+public class MostViewedMacroViewPages extends ViewPage
```
Missing JavaDoc explaining what this page object represents (e.g. which UI element). Also, the class name is very confusing, but I can't suggest a better name because without a JavaDoc it's hard to understand what this page object is for.

mflorea commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/MostViewedMacroViewPages.java:
```diff
@@ -0,0 +1,71 @@
+        driver =  getUtil().getDriver();
+    }
+
+    public static MostViewedMacroViewPages gotoPage()
```
Why is this and the rest of the following methods static? A page object is not a "Utils" class providing static utility methods for tests. We use page objects to model in an object-oriented way the UI that the user interacts with. The web page the user interacts with may have a header (top bar), a menu, a body, a footer. The body could include a form, which includes multiple fields. Two of these fields could be a rich text editor, which has a toolbar and a rich text area, so on. All these UI elements are usually modeled by separate page objects, allowing you to write tests that do:

    new RichTextEditor(somePage.getForm('foo').getField('bar')).getToolBar().click('Bold');

mflorea commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/MostViewedMacroViewPages.java:
```diff
@@ -0,0 +1,71 @@
+        return new MostViewedMacroViewPages();
+    }
+
+    public static void openRowEvolutionModal()
+    {
+        driver.waitUntilElementIsVisible(By.cssSelector(ROW_EVOLUTION_BUTTON_SELECTOR));
+        driver.findElement(By.cssSelector(ROW_EVOLUTION_BUTTON_SELECTOR)).click();
+    }
```
You should return the modal, so that the test can interact with it.

mflorea commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/MostViewedMacroViewPages.java:
```diff
@@ -0,0 +1,71 @@
+        return driver.findElement(By.cssSelector(".xcontent h2 div a")).getAttribute("title");
+
+    }
+    public static void closeModal()
+    {
+        driver.findElement(By.cssSelector(".btn.btn-default")).click();
+    }
```
This is already provided by ``BaseModal``. No need to reimplement the action.

mflorea commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/MostViewedMacroViewPages.java:
```diff
@@ -0,0 +1,71 @@
+        driver.findElement(By.cssSelector(ROW_EVOLUTION_BUTTON_SELECTOR)).click();
+    }
+
+    public static boolean isModalDisplayed()
+    {
+        driver.waitUntilElementIsVisible(By.cssSelector(ROW_EVOLUTION_MODAL_SELECTOR));
+        if(driver.findElement(By.cssSelector(ROW_EVOLUTION_MODAL_SELECTOR)).isDisplayed())
+            return true;
+        return false;
+    }
```
Based on the method name you would never guess that it waits for the modal to be visible. And then, if it waits for the modal to be visible, isn't it always going to be visible after the wait ends without a timeout exception? Moreover, the wait needs to be done by the method that opens the modal. In general, a page object method that triggers an action should wait (by default) for that action to end (it can provide an alternative signature to skip the wait if needed).

mflorea commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/MostViewedMacroViewPages.java:
```diff
@@ -0,0 +1,71 @@
+            return true;
+        return false;
+    }
+        public static String getMacroDescription()
+    {
+        driver.waitUntilElementIsVisible(By.cssSelector(".xcontent h2 div a"));
+        return driver.findElement(By.cssSelector(".xcontent h2 div a")).getAttribute("title");
+
+    }
```
Same comment as above. The method name doesn't give any clue that it waits until the macro description is visible. In general, getter page object methods shouldn't do any wait (besides that default wait performed by ``findElement``). The wait is done by the method that triggered the action for which we want to get the outcome.

mflorea commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/MatomoViewPage.java:
```diff
@@ -0,0 +1,57 @@
+
+import static org.junit.Assert.fail;
+
+public class MatomoViewPage extends ViewPage
```
Missing JavaDoc.

mflorea commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/MatomoViewPage.java:
Please format the code.

mflorea commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/MatomoViewPage.java:
```diff
@@ -0,0 +1,57 @@
+
+    static public void returnToXwiki()
+    {
+        getUtil().gotoPage("http://localhost:8080/xwiki/bin/view/Main/");
```
You shouldn't hard-code the URL used to access XWiki because this can be different depending on the environment where the tests are run (e.g. inside a Docker container). There is API to create the URL.

mflorea commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/HomePageViewPage.java:
```diff
@@ -0,0 +1,116 @@
+        DocumentReference documentReference = new DocumentReference("xwiki", "Analytics", "WebHome");
+        Map<String, String> params = new HashMap<>();
+        params.put("force", "1");
+        getUtil().gotoPage(documentReference, "edit", params);
+        return new HomePageViewPage();
```
You're going to edit mode, but you return a "*ViewPage"? Doesn't look right.

mflorea commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/MostViewedMacroViewPages.java:
```diff
@@ -0,0 +1,71 @@
+    public static boolean isModalDisplayed()
+    {
+        driver.waitUntilElementIsVisible(By.cssSelector(ROW_EVOLUTION_MODAL_SELECTOR));
+        if(driver.findElement(By.cssSelector(ROW_EVOLUTION_MODAL_SELECTOR)).isDisplayed())
+            return true;
```
Always wrap the body of an IF in a block ``{...}``.

mflorea commented on application-analytics-test/application-analytics-test-docker/src/test/it/xwiki/analytics/test/ui/config/Config.java:
```diff
@@ -0,0 +1,59 @@
+
+public class Config
+{
+    public static final String DB_CONTAINER_NAME = "farcasut/custom-mysql:latest";
```
The test needs to run on CI. I doubt this custom Docker container is published somewhere, but even if it is, we should use an official container.

Farcasut replied:
The container is currently publicly hosted on Docker Hub, but do you think it would be a better choice to create the container from scratch? Like using the default container and populating the database using the API provided by TestContainers

mflorea commented on application-analytics-test/application-analytics-test-docker/src/test/it/xwiki/analytics/test/ui/config/Config.java:
```diff
@@ -0,0 +1,59 @@
+
+    public static final String MATOMO_CREDENTIALS = "ADMIN1";
+
+    //139fe0f667fe8eb309cad6179e486d1e
```
Leftover?

mflorea commented on application-analytics-test/application-analytics-test-docker/src/test/it/xwiki/analytics/test/ui/config/Config.java:
```diff
@@ -0,0 +1,59 @@
+
+    public static String getTrackingCode()
+    {
+        return "<!-- Matomo -->\n" + "<script>\n" + "  var _paq = window._paq = window._paq || [];\n"
+            + "  /* tracker methods like \"setCustomDimension\" should be called before \"trackPageView\" */\n"
+            + "  _paq.push(['trackPageView']);\n" + "  _paq.push(['enableLinkTracking']);\n" + "  (function() {\n"
+            + "    var u=\"http://" + Config.ADDRESS + ":" + Config.MATOMO_BRIDGE_PORT + "/" + "\";\n"
+            + "    _paq.push(['setTrackerUrl', u+'matomo.php']);\n" + "    _paq.push(['setSiteId', '1']);\n"
+            + "    var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];\n"
+            + "    g.async=true; g.src=u+'36011373.js'; s.parentNode.insertBefore(g,s);\n" + "  })();\n" + "</script>\n"
+            + "<!-- End Matomo Code -->\n";
```
Either use ``StringBuilder`` to format this code properly or put the code in a separate file that you read here.

mflorea commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/AdminViewPage.java:
```diff
@@ -0,0 +1,123 @@
+        return this;
+    }
+
+    public boolean inProgressNotification(String message)
+    {
+
+        try {
+            this.waitForNotificationInProgressMessage(message);
+            return true;
+        } catch (Exception e) {
+            return false;
+        }
+    }
+
+    public boolean errorNotification(String message)
+    {
+
+        try {
+            this.waitForNotificationErrorMessage(message);
+            return true;
+        } catch (Exception e) {
+            return false;
+        }
+    }
+
+    public boolean successNotification(String message)
+    {
+        try {
+            this.waitForNotificationSuccessMessage(message);
+            return true;
+        } catch (Exception e) {
+            return false;
+        }
+    }
```
I don't understand the need for these methods. If the wait throws an exception then the test fails. Why write

     assertTrue(adminViewPage.successNotification("Saved"));

when you can simply write:

    adminViewPage.waitForNotificationSuccessMessage("Saved");

The test will continue if the wait is successful, otherwise it will fail with a timeout.

mflorea reviewed (changes requested):
Your code needs significant refactoring. I recommend checking the page objects and tests from XWiki platform to take inspiration from.

Farcasut committed:
Create functional tests #45
*
-Update the poms.xml
-Moved from a class with static methods to a normal one

Farcasut committed:
Create functional tests #45
*
- More comments

oanalavinia commented on application-analytics-test/application-analytics-test-docker/src/test/it/xwiki/analytics/test/ui/AnalyticsIT.java:
```diff
@@ -0,0 +1,224 @@
+        adminViewPage.setTrackingCode("").setAuthTokenId(Config.MATOMO_AUTH_TOKEN).setIdSiteId("1")
+            .setRequestAddressId(Config.ADDRESS + ":" + Config.MATOMO_BRIDGE_PORT).bringSaveButtonIntoView();
+
+        adminViewPage.inProgressNotification("Saving...");
```
Why not use directly `adminViewPage.waitForNotificationInProgressMessage("Saving...");` ? The name is also more suggestive

oanalavinia commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/HomePageEditPage.java:
```diff
@@ -0,0 +1,99 @@
+    public HomePageViewPage saveDashboard()
+    {
+
+        EditPage editPage = new EditPage();
+        editPage.clickSaveAndView();
```
`this` is already an EditPage, why do you need a new object?
For me it looks like you don't actually need this method, since you can directly do the `myHomePageEditPage.clickSaveAndView()` from where the method is called. If after you will need the HomePageViewPage, just use `new HomePageViewPage()`

oanalavinia commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/HomePageEditPage.java:
```diff
@@ -0,0 +1,99 @@
+import org.xwiki.test.ui.XWikiWebDriver;
+import org.xwiki.test.ui.po.editor.EditPage;
+
+public class HomePageEditPage extends EditPage
```
The name could be more specific. Maybe `AnalyticsEditPage`? Same for ViewPage

oanalavinia commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/AdminViewPage.java:
```diff
@@ -0,0 +1,107 @@
+import org.xwiki.test.ui.XWikiWebDriver;
+import org.xwiki.test.ui.po.ViewPage;
+
+public class AdminViewPage extends ViewPage
```
The name could be more specific, maybe `AnalyticsConfigViewPage`?

oanalavinia commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/HomePageEditPage.java:
```diff
@@ -0,0 +1,99 @@
+        driver.findElement(By.cssSelector(css)).click();
+    }
+
+    private void selectMacro(String macroID, String macroName)
```
I see that starting with XWiki 14.9 some API for what you are doing here was added https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/MacroDialogSelectModal.java
I'm wandering if it would be good to use the 14.10 parent and benefit from this? We will be moving to this parent pretty soon anyway. If not, I still feel that we could duplicate the `MacroDialogSelectModal` class and remove it once the parent is upgraded. @mflorea @Farcasut WDYT?

oanalavinia commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/MostViewedPagesMacroViewPage.java:
```diff
@@ -0,0 +1,75 @@
+     */
+    public String getMacroDescription()
+    {
+        driver.waitUntilElementIsVisible(By.cssSelector(".xcontent h2 div a"));
```
This seems really fragile, maybe you could add a class to the element?

oanalavinia commented on application-analytics-test/application-analytics-test-docker/src/test/it/xwiki/analytics/test/ui/AnalyticsIT.java:
```diff
@@ -0,0 +1,224 @@
+    {
+        MostViewedPagesMacroViewPage mostViewedPagesMacroViewPage = new MostViewedPagesMacroViewPage();
+        BaseModal baseModal = mostViewedPagesMacroViewPage.gotoPage().openRowEvolutionModal();
+        System.out.println("/MODAL/ " + baseModal.isDisplayed() + "/A/A");
```
Leftover?

oanalavinia commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/MostViewedPagesMacroViewPage.java:
```diff
@@ -0,0 +1,75 @@
+     */
+    public BaseModal openRowEvolutionModal()
+    {
+        BaseModal baseModal = new BaseModal(By.cssSelector(".modal-dialog.modal-lg"));
```
You should create a PO for the row evolution modal (check https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-test/xwiki-platform-test-ui/src/main/java/org/xwiki/test/ui/po/ConfirmationModal.java )

oanalavinia commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/HomePageEditPage.java:
```diff
@@ -0,0 +1,99 @@
+     * @return
+     */
+
+    public HomePageEditPage removeLastMacro()
```
This method seems fragile. Anyway, it should be part of the https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-test/xwiki-platform-ckeditor-test-pageobjects/src/main/java/org/xwiki/ckeditor/test/po/MacroDialogSelectModal.java API. If you added it only for the edit permission check, I feel that it's safer to only test the add macro usecase. WDYT?

mflorea commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/AdminViewPage.java:
```diff
@@ -0,0 +1,107 @@
+
+    public AdminViewPage setTrackingCode(String value)
+    {
+        WebElement element = driver.findElement(By.id(TRACKING_CODE_ID));
```
Check existing page objects. You can use the`` @FindBy`` annotation on ``WebElement`` fields of a class to inject the value so that you don't need to look up the elements yourself, unless this element is displayed dynamically after the page loads.

Farcasut committed:
Updated the parent to 14.10 and added new dependencies to make sure that everything works properly

Farcasut commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/AdminViewPage.java:
```diff
@@ -0,0 +1,107 @@
+        return this;
+    }
+
+    public AdminViewPage bringSaveButtonIntoView()
```
Yes, the main action of this method is a click that saves the configs. I think a better name would be saveConfigs

Farcasut commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/AdminViewPage.java:
```diff
@@ -0,0 +1,107 @@
+        return this;
+    }
+
+    public void inProgressNotification(String message)
+    {
+        this.waitForNotificationInProgressMessage(message);
+    }
+
+    public void errorNotification(String message)
+    {
+        this.waitForNotificationErrorMessage(message);
+    }
+
+    public void successNotification(String message)
+    {
+        this.waitForNotificationSuccessMessage(message);
+    }
```
It was a mistake. I had a bigger method, but I realized that it was redundant and forgot to remove the methods completely.

Farcasut committed:
Renamed classes and methods, added a class to a html element to make it easier to select it.

Farcasut committed:
Create functional tests #45
Added a FindBy to select the elements in the config tab
Removed unnecessary imports
Renamed classes
Modified the URL of the rowevolution request to have the get action insted of the view

Farcasut commented on application-analytics-test/application-analytics-test-docker/pom.xml:
```diff
@@ -0,0 +1,125 @@
+    <!-- I need the Menu Application to test the custom configs. -->
+    <dependency>
+      <groupId>org.xwiki.platform</groupId>
+      <artifactId>xwiki-platform-menu-ui</artifactId>
```
I've tried using the administration instead of the menu, but I get an error when I try to access the config menu.

Farcasut committed:
Create functional tests #45
Added a wait to avoid flickering

Farcasut committed:
Merge branch 'main' into functionalTests

oanalavinia commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/RowEvolutionModal.java:
```diff
@@ -0,0 +1,39 @@
+    public RowEvolutionModal(By selector) {
+        super(selector);
+    }
+    public RowEvolutionModal openModal()
```
You should have a method in `MostViewdPagesElement` that opens the modal and has RowEvolutionModal as the return type
You can check the [EditGroupModal](https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-administration/xwiki-platform-administration-test/xwiki-platform-administration-test-pageobjects/src/main/java/org/xwiki/administration/test/po/EditGroupModal.java) exampe and how is it used in [GroupsPage](https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-administration/xwiki-platform-administration-test/xwiki-platform-administration-test-pageobjects/src/main/java/org/xwiki/administration/test/po/GroupsPage.java#L107)

oanalavinia commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/MostViewedPagesElement.java:
```diff
@@ -0,0 +1,57 @@
+    WebElement macroOutput;
+    public MostViewedPagesElement(String id)
+    {
+        getUtil().getDriver().waitUntilElementIsVisible(By.id(id));
```
It would be cleaner to move this inside a waitUntilReady method

oanalavinia commented on application-analytics-test/application-analytics-test-pageobjects/src/main/java/com/xwiki/analytics/test/po/MostViewedPagesElement.java:
```diff
@@ -0,0 +1,57 @@
+{
+
+
+    WebElement macroOutput;
```
This should be private. You can also rename it to `container`, WDYT?

Farcasut committed:
Create functional tests #45
Refactored code, added a new test and improved formatted the code.
